llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

```diff
@@ -217,10 +217,20 @@ bool SIInstrInfo::isSafeToSink(MachineInstr &MI,
         SmallVector<MachineBasicBlock *, 1> ExitBlocks;
         FromCycle->getExitBlocks(ExitBlocks);
         assert(ExitBlocks.size() == 1);
-        assert(ExitBlocks[0]->getSinglePredecessor());
-
+        // After structurize-cfg, cycle exit block should have exactly one
```

Comment:
Isn't it simpler to write this code so that it makes no assumptions about structurized CFGs? E.g.
```c++
  FromCycle->getExitBlocks(ExitBlocks);
  for (ExitBlock : ExitBlocks)
    for (Pred : ExitBlock->predecessors())
      if (FromCycle->contains(Pred) && hasDivergentBranch(Pred))
        return false;
```

Reply:
OK, write it like that for simplicity and delete asserts. Can we leave some of the comments? I don't like fully hiding expected CFG as it could give wrong impression that we can deal with cycle with two divergent exits or similar

Reply:
> it could give wrong impression that we can deal with cycle with two divergent exits or similar

*This* code can deal with all those cases, can't it?

Reply:
Your proposal without asserts would return something even if input was not "structurized-cfg". Do we want to return something even when CFG is something we don't handle?

Reply:
Yes, if *this* code can easily handle arbitrary CFGs then it should do so.

If other parts of the code cannot handle arbitrary CFGs, then they should detect the problems.

Reply:
Yes, whether we return a block should be independent of structrize-cfg. But we should keep the comment for our own future help.